detect-python-version/action.yml

```diff
@@ -0,0 +1,42 @@
+name: Determine Python and Poetry versions
+description: Determine Python and Poetry versions
+
+inputs:
+  python_version:
+    description: "Python version to use"
+    required: false
+  working_directory:
+    description: "Working directory, defaults to GITHUB_WORKSPACE"
+    required: false
+    default: ${{ github.workspace }}
+
+outputs:
+  python-major-version:
+    description: "Python version"
+    value: ${{ steps.detect-versions.outputs.python-major-version }}
+  poetry-version:
+    description: "Poetry version"
+    value: ${{ steps.detect-versions.outputs.poetry-version }}
+
+runs:
+  using: composite
+  steps:
+    # Empty arrays cannot be provided with `-f`, see https://github.com/cli/cli/issues/1484
+    - id: detect-versions
+      run: |
+        if [ -z "${{ inputs.python_version }}" ] ; then
+            PYTHON_VERSION="$(sed -n -e '/^\[metadata\]/,/^\[/p' poetry.lock | sed -n -e 's/^python-versions[[:space:]]*=[[:space:]]*//p' | tr -d \"'[:space:]'\'~^)"
+        else
+            PYTHON_VERSION="${{ inputs.python_version }}"
+        fi
+
+        if [ -z "${{ inputs.poetry_version }}" ] ; then
+            POETRY_VERSION="$(head --lines=1 poetry.lock | grep --only-matching '[0-9]\+\.[0-9]\+\.[0-9]\+')"
+        else
+            POETRY_VERIONS="${{ inputs.poetry_version }}"
+        fi
+
+        echo "python-major-version=$PYTHON_VERSION" >> $GITHUB_OUTPUT
```

Comment:
As discussed, this Python version is the major version something like `3.12` in our use case, but in theory it can be 3.12.0 if you specify this in `pyproject.toml` instead of `~3.12`. This could be confusing.

Do we really want to update this @IamAbbey @zyv. For me both is okay, just wanted to mention this.

Reply:
I am okay by it, since for our specific use case, its something like `~3.12`

Reply:
For me all suggestions are fine, just wanted to bring this point up, that in theory `python-major-version` is not correct. So I thought it is okay to leave it as it is with `python-version`. 
But as I mentioned, `python-major-version` or `python-version-constraint` would be also fine.

@felix11h Do you have a strong opinion on that? Otherwise, @IamAbbey please decide what you prefer.

Reply:
The way I know semantic versioning is of the form `MAJOR.MINOR.PATCH`, see for example https://semver.org/. So, `python-major-version` seems a bit confusing to me, as I understand it is the goal to refer to, for example, `3.12` instead of just `3`.

I like `python-version`, which is consistent with the way action/setup-python is using it, or alternatively, `python-version-constraint`. Maybe the best way could be to go with `python-version`, but add a more detailed description to the output variable. Something like:

```yaml
    description: "Python version. Can be exact version (3.12.0) or version constraint (3.12)"
```

Reply:
Thanks for your contribution @felix11h